modules/core/src/test/java/org/apache/ignite/internal/processors/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,426 @@
+
+        cfg.setAtomicConfiguration(atomicCfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        startGridsMultiThreaded(NODES_CNT);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * Test a creating of the reentrant lock.
+     *
+     * @throws Exception If failed.
+     */
+    public void testInitialization() throws Exception {
+        try (IgniteLock lock = createReentrantLock(0, "lock1", false)) {
+        }
+
+        try (IgniteLock lock = createReentrantLock(0, "lock2", true)) {
+        }
+    }
+
+    /**
+     * Test a sequential lock acquiring and releasing.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock() throws Exception {
+        testReentrantLock(true);
+        testReentrantLock(false);
+    }
+
+    /**
+     * {@link this#testReentrantLock()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock(boolean fair) throws Exception {
+        for (int i = 0; i < NODES_CNT; i++) {
+            final IgniteLock lock = createReentrantLock(i, fair ? "lock1" : "lock2", fair);
+
+            // Other node can still don't see update.
+            GridTestUtils.waitForCondition(new GridAbsPredicate() {
+                @Override public boolean apply() {
+                    return !lock.isLocked();
+                }
+            }, 500L);
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.unlock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.unlock();
+
+            assertFalse(lock.isLocked());
+            assertFalse(lock.isHeldByCurrentThread());
+        }
+    }
+
+    /**
+     * Test an async lock acquiring and releasing with many nodes.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode() throws Exception {
+        testReentrantLockMultinode(false);
+        testReentrantLockMultinode(true);
+    }
+
+    /**
+     * {@link this#testReentrantLockMultinode()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode(final boolean fair) throws Exception {
+        final AtomicInteger inx = new AtomicInteger(0);
+
+        GridTestUtils.runMultiThreadedAsync(new Runnable() {
+            @Override public void run() {
+                try {
+                    IgniteLock lock = createReentrantLock(inx.getAndIncrement(), fair ? "lock1" : "lock2", fair);
+                    lock.lock();
+
+                    try {
+                        assertTrue(lock.isLocked());
```

Comment:
newline mb

Reply:
Where? after try?

Reply:
newline between assertTrue(lock.isLocked()); and Thread.sleep(1_000L);

Reply:
fixed